rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -178,29 +168,32 @@ class ReplayRSpace[F[_]: Concurrent: ContextShift: Log: Metrics: Span, C, P, A,
               comm.consume == source
           }.toSeq),
       c =>
-        store
-          .getData(c)
-          .map(_.iterator.zipWithIndex.toSeq)
-          .map(
-            as => {
-              c -> {
-                if (c == channel)
-                  (Datum(data, persist, produceRef), -1) +: as
-                else as
-              }.filter {
-                matches(comm)
-              }
-            }
-          )
+        for {
+          seqDatum          <- store.getData(c)
+          datumWithIndexSeq = seqDatum.iterator.zipWithIndex.toSeq
+          extendedDatumWithIndexSeq = if (c == channel)
+            (Datum(data, persist, produceRef), -1) +: datumWithIndexSeq
+          else datumWithIndexSeq
+          filteredExtendedDatumWithIndexSeq <- extendedDatumWithIndexSeq.toList.filterA(
+                                                matches(comm)
+                                              )
+        } yield (c, filteredExtendedDatumWithIndexSeq)
```

Comment:
This code now seems much more noisy with for..yield syntax, it's more difficult to follow with many variables.
```scala
store
  .getData(c)
  // Convert to immutable.Seq because of Traversable instance defined
  .map(_.iterator.zipWithIndex.to[immutable.Seq])
  .flatMap { datums =>
    (if (c == channel) (Datum(data, persist, produceRef), -1) +: datums else datums)
      .filterA(matches(comm))
      .map(_.to[Seq])
  }
  .map((c, _))
```
It also contains one subtle difference with conversion to `immutable.Seq` which has `filterA` method defined. This is critical path of the code and conversion to List introduces intermediate structure which can be avoided.

More about Traversable instance for immutable.Seq. https://github.com/typelevel/cats/pull/3620